src/main/java/hunternif/mc/atlas/network/client/TileNameIDPacket.java

```diff
@@ -4,83 +4,80 @@
+            // Reading negative value to save on traffic, because custom biome
+            // IDs are always negative.
+            int biomeID = -buffer.readVarInt();
+            nameToIdMap_.put(new ResourceLocation(name), biomeID);
+        }
+        return new TileNameIDPacket(nameToIdMap_);
+    }
 
-	@Override
-	public void write(PacketByteBuf buffer) throws IOException {
-		buffer.writeVarInt(nameToIdMap.size());
-		for (Entry<Identifier, Integer> entry : nameToIdMap.entrySet()) {
-			buffer.writeString(entry.getKey().toString());
-			// Writing negative value to save on traffic, because custom biome
-			// IDs are always negative.
-			buffer.writeVarInt(-entry.getValue());
-		}
-	}
+    public static void write(TileNameIDPacket msg, PacketBuffer buffer) {
+        buffer.writeVarInt(msg.nameToIdMap.size());
+        for (Entry<ResourceLocation, Integer> entry : msg.nameToIdMap.entrySet()) {
+            buffer.writeString(entry.getKey().toString());
+            // Writing negative value to save on traffic, because custom biome
+            // IDs are always negative.
+            buffer.writeVarInt(-entry.getValue());
+        }
+    }
 
-	@Override
-	@Environment(EnvType.CLIENT)
-	protected void process(PlayerEntity player, EnvType side) {
-		for (Entry<Identifier, Integer> entry : nameToIdMap.entrySet()) {
-			Identifier tileName = entry.getKey();
-			int id = entry.getValue();
-			// Remove old texture mapping
-			int oldID = ExtTileIdMap.instance().getPseudoBiomeID(tileName);
-			if (oldID != ExtTileIdMap.NOT_FOUND && oldID != id) {
-				BiomeTextureMap.instance().setTexture(Registry.BIOME.get(oldID), null);
-			}
-			ExtTileIdMap.instance().setPseudoBiomeID(tileName, id);
-			// Register new texture mapping
-			TextureSet texture = ExtTileTextureMap.instance().getTexture(tileName);
-			BiomeTextureMap.instance().setTexture(id, texture);
-		}
-		TileIdRegisteredCallback.EVENT.invoker().onTileIDsReceived(nameToIdMap);
-	}
+    public void process(Supplier<NetworkEvent.Context> ctx) {
+        ctx.get().enqueueWork(() -> {
+            for (Entry<ResourceLocation, Integer> entry : nameToIdMap.entrySet()) {
+                ResourceLocation tileName = entry.getKey();
+                int id = entry.getValue();
+                // Remove old texture mapping
+                int oldID = ExtTileIdMap.instance().getPseudoBiomeID(tileName);
+                if (oldID != ExtTileIdMap.NOT_FOUND && oldID != id) {
+                    BiomeTextureMap.instance().setTexture(Registry.BIOME.getByValue(oldID), null);
+                }
+                ExtTileIdMap.instance().setPseudoBiomeID(tileName, id);
+                // Register new texture mapping
+                TextureSet texture = ExtTileTextureMap.instance().getTexture(tileName);
+                BiomeTextureMap.instance().setTexture(id, texture);
+            }
+            //TileIdRegisteredEvent.EVENT.invoker().onTileIDsReceived(nameToIdMap);

```

Comment:
TODO: check if TileIdRegisteredEvent is necessary, and uncomment or delete this line and other related lines

Reply:
It is probably neccessary for the API, but I didn't have a good way of doing it so just left it out.